lib/matplotlib/__init__.py

```diff
@@ -761,30 +763,85 @@ def __getitem__(self, key):
 
         return self._get(key)
 
+    def get_default(self, key):
+        """Return default value for the key set during initialization."""
+        if key in _deprecated_map:
+            version, alt_key, alt_val, inverse_alt = _deprecated_map[key]
+            _api.warn_deprecated(
+                version, name=key, obj_type="rcparam", alternative=alt_key)
+            return inverse_alt(self._get(alt_key))
+
+        elif key in _deprecated_ignore_map:
+            version, alt_key = _deprecated_ignore_map[key]
+            _api.warn_deprecated(
+                version, name=key, obj_type="rcparam", alternative=alt_key)
+            return self._defaults[alt_key] if alt_key else None
+
+        return self._defaults[key]
+
+    def get_defaults(self):
+        """Return default values set during initialization."""
+        return self._defaults.copy()
+
     def _get_backend_or_none(self):
         """Get the requested backend, if any, without triggering resolution."""
         backend = self._get("backend")
         return None if backend is rcsetup._auto_backend_sentinel else backend
 
+    def __delitem__(self, key):
+        if key not in self.validate:
+            raise KeyError(
+                f"{key} is not a valid rc parameter (see rcParams.keys() for "
+                f"a list of valid parameters)")
+        try:
+            del self._rcvalues[key]
+        except KeyError as err:
+            raise KeyError(
+                f"No custom value set for {key}. Cannot delete default value."
+            ) from err
+
+    def __contains__(self, key):
+        return key in self._rcvalues
+
+    def __iter__(self):
+        """Yield from sorted list of keys"""
+        with _api.suppress_matplotlib_deprecation_warning():
+            yield from sorted(self._rcvalues.keys())
+
+    def __len__(self):
+        return len(self._rcvalues)
+
     def __repr__(self):
         class_name = self.__class__.__name__
         indent = len(class_name) + 1
         with _api.suppress_matplotlib_deprecation_warning():
-            repr_split = pprint.pformat(dict(self), indent=1,
+            repr_split = pprint.pformat(dict(self._rcvalues.items()), indent=1,
                                         width=80 - indent).split('\n')
         repr_indented = ('\n' + ' ' * indent).join(repr_split)
         return f'{class_name}({repr_indented})'
 
     def __str__(self):
-        return '\n'.join(map('{0[0]}: {0[1]}'.format, sorted(self.items())))
+        return '\n'.join(map('{0[0]}: {0[1]}'.format, sorted(self._rcvalues.items())))
 
-    def __iter__(self):
-        """Yield sorted list of keys."""
-        with _api.suppress_matplotlib_deprecation_warning():
-            yield from sorted(dict.__iter__(self))
+    @_api.deprecated("3.8")
+    def clear(self):
+        pass
```

Comment:
On one hand, `clear` is current a no-op because `popitem` raises `KeyError` which makes the default implementation from `MutableMapping` conclude that there is nothing to do and it is cleared despite doing nothing.

This is part of the public API of the `MutableMapping` API so I don't think we can (should) actually deprecate this.

We should either make sure the note says "we are going to make this error in the future" or just warn "this is not doing what you think it is doing".